feat: improve RSC content detection in export mode when Content-Type …

Vansh5632 opened the pull request:
## For Contributors

### Fixing a bug / Adding a feature
#82451 

**Checklist:**

- [x] **Related issues linked using `fixes #82451 `**
  - This addresses issues where CDNs/proxies strip Content-Type headers in export mode, causing unnecessary MPA navigation fallbacks
  - *Note: If there's a specific GitHub issue for this problem, reference it with `fixes #82451 `*

- [x] **Tests added**
  - Comprehensive unit tests in `packages/next/src/client/components/router-reducer/tests/fetch-server-response.test.ts`
  - Tests cover URL processing utilities, RSC payload detection heuristics, edge cases, and environment configurations
  - 9 passing test cases covering all major scenarios

- [x] **Documentation added**
  - Inline code documentation explaining the sniffing logic
  - JSDoc comments for new functions (`peekAndSniff`, `detectRscOrTextPayload`)
  - Detailed commit message explaining the feature

- [x] **Feature follows contribution guidelines**
  - Performance-conscious implementation (512-byte peek limit)
  - Conservative heuristics to avoid false positives
  - Only activates in specific conditions (production + export mode + missing Content-Type)
  - Maintains backward compatibility
## Video

https://github.com/user-attachments/assets/857c209b-e8d4-4032-be50-97d798622960



## Pull Request Description

### What?
Improves RSC content detection in production export mode when Content-Type headers are missing or incorrect by adding intelligent content sniffing.

### Why?
In production with `output: 'export'`, CDNs and proxies sometimes strip or modify Content-Type headers, causing valid RSC prefetch responses to be incorrectly rejected and fall back to MPA navigation unnecessarily. This reduces performance and user experience.

### How?
- **Content sniffing**: Peeks at the first 512 bytes of response streams using `stream.tee()`
- **RSC pattern detection**: Recognizes common RSC payload markers like `0:[`, `1:"`, etc.
- **Text validation**: Ensures content is mostly printable ASCII (≥90%) without HTML markers
- **HTML rejection**: Explicitly rejects content starting with `<!DOCTYPE`, `<html>`
- **Conservative approach**: Only activates when specific conditions are met
- **Stream integrity**: Uses `tee()` to avoid consuming the main response body

**Files Changed:**
- `packages/next/src/client/components/router-reducer/fetch-server-response.ts` - Main implementation
- `packages/next/src/client/components/router-reducer/tests/fetch-server-response.test.ts` - Test coverage

**Key Features:**
- Performance conscious (512-byte peek limit)
- Safety first (conservative heuristics)
- Non-breaking (only affects edge cases in export mode)
- Well tested (9 comprehensive test cases)

**Test Coverage:**
```typescript
describe('RSC payload detection heuristics', () => {
  // Tests for RSC pattern detection
  it('should detect RSC payload patterns', () => {
    expect(detectRscOrTextPayload('0:["test"]')).toBe(true)
    expect(detectRscOrTextPayload('1:"hello world"')).toBe(true)
    expect(detectRscOrTextPayload('"children":["value"]')).toBe(true)
  })

  // Tests for valid text content
  it('should detect valid text content', () => {
    expect(detectRscOrTextPayload('This is valid text content')).toBe(true)
    expect(detectRscOrTextPayload('Multi-line\ntext\ncontent')).toBe(true)
  })

  // Tests for HTML rejection
  it('should reject HTML content', () => {
    expect(detectRscOrTextPayload('<!DOCTYPE html>')).toBe(false)
    expect(detectRscOrTextPayload('<html><head>')).toBe(false)
  })

  // Tests for edge cases
  it('should handle mixed content appropriately', () => {
    const binaryLikeContent = 'abc\0\0\0def\0\0\0'
    expect(detectRscOrTextPayload(binaryLikeContent)).toBe(false)

    const mostlyPrintableContent = 'This is mostly good content with some \0 binary'
    expect(detectRscOrTextPayload(mostlyPrintableContent)).toBe(true)
  })
})

Vansh5632 committed:
feat: improve RSC content detection in export mode when Content-Type is missing

- Add content sniffing for valid RSC payloads in production export mode
- Peek at response stream prefix (512 bytes) to detect RSC patterns
- Use stream.tee() to avoid consuming the main response body
- Apply conservative heuristics to reject HTML and accept valid text/RSC
- Only activate when Content-Type is missing/incorrect in export mode
- Add comprehensive test coverage for detection logic
- Maintain backward compatibility and performance

Fixes cases where CDNs/proxies strip Content-Type headers causing
unnecessary MPA navigation fallbacks in static export deployments.

ijjk commented:
Allow CI Workflow Run

- [ ] approve CI run for commit: eb36c380498d65af91590168ac55146b6d31363c

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

Vansh5632 committed:
fix: properly maintain Response prototype when replacing body stream

- Replace spread operator with explicit property assignment
- Use 'satisfies RSCResponse' to ensure type safety
- Preserves Headers prototype and all Response methods
- Fixes runtime failures in downstream code that expects proper Response interface

Resolves bot feedback about prototype loss and missing properties.
